File: pwa/js/lookup.js
// 英単語を引く（Mac 版 lookup.py と同じ順）。
//   1. EJDict（dict/<頭文字>.json）をそのまま  2. 活用を戻す  3. 派生語を戻す（近い語）  4. ハイフンの部分
//   5. Free Dictionary API（英英。ネットがあるときだけ）
const shards = new Map();

async function shard(ch) {
  if (!shards.has(ch)) {
    shards.set(ch, fetch(`dict/${ch}.json`).then((r) => (r.ok ? r.json() : {})).catch(() => ({})));
  }
  return shards.get(ch);
}

async function ej(w) {
  if (!/^[a-z]/.test(w)) return [];
  const d = await shard(w[0]);
  return (d[w] || []).slice(0, 5).map(([word, mean]) => ({ word, mean }));
}

export function normalize(word) {
  return word.trim().replace(/^[.,;:!?()[\]{}"“”‘’']+|[.,;:!?()[\]{}"“”‘’']+$/g, "").toLowerCase().replace(/['’]s$/, "");
}

export function inflectionCandidates(w) {
  const c = [];
  const add = (x) => { if (x.length >= 2 && !c.includes(x) && x !== w) c.push(x); };
  if (w.endsWith("ies")) add(w.slice(0, -3) + "y");
  if (w.endsWith("es")) add(w.slice(0, -2));
  if (w.endsWith("s") && !w.endsWith("ss")) add(w.slice(0, -1));
  if (w.endsWith("ied")) add(w.slice(0, -3) + "y");
  if (w.endsWith("ed")) {
    add(w.slice(0, -2)); add(w.slice(0, -1));
    if (w.length > 4 && w.at(-3) === w.at(-4)) add(w.slice(0, -3));
  }
  if (w.endsWith("ing")) {
    add(w.slice(0, -3)); add(w.slice(0, -3) + "e");
    if (w.length > 5 && w.at(-4) === w.at(-5)) add(w.slice(0, -4));
  }
  if (w.endsWith("er")) { add(w.slice(0, -2)); add(w.slice(0, -1)); }
  if (w.endsWith("est")) { add(w.slice(0, -3)); add(w.slice(0, -2)); }
  if (w.endsWith("ly")) add(w.slice(0, -2));
  return c;
}

const DERIVATIONS = [["ability", ""], ["ability", "e"], ["able", ""], ["able", "e"], ["ible", ""], ["ible", "e"],
  ["ator", "ate"], ["ators", "ate"], ["or", ""], ["ors", ""], ["ation", "ate"], ["ation", "e"], ["ations", "ate"],
  ["ation", ""], ["ment", ""], ["ments", ""], ["ness", ""], ["ity", ""], ["ity", "e"], ["ization", "ize"],
  ["ized", "ize"], ["ically", "ic"], ["ically", "ical"]];

export function derivationCandidates(w) {
  const out = [];
  for (const [suf, rep] of DERIVATIONS) {
    if (w.endsWith(suf) && w.length - suf.length >= 3) {
      const x = w.slice(0, -suf.length) + rep;
      if (!out.includes(x) && x !== w) out.push(x);
    }
  }
  return out;
}

async function api(w) {
  if (!navigator.onLine) return [];
  try {
    const r = await fetch(`https://api.dictionaryapi.dev/api/v2/entries/en/${encodeURIComponent(w)}`);
    if (!r.ok) return [];
    const e = (await r.json())[0];
    const parts = (e.meanings || []).slice(0, 3).map((m) =>
      `(${m.partOfSpeech || ""}) ` + (m.definitions || []).slice(0, 2).map((d) => d.definition).join(" / "));
    return parts.length ? [{ word: e.word || w, mean: parts.join(" / ") }] : [];
  } catch { return []; }
}

export async function lookup(word) {
  const q = normalize(word);
  const res = { query: word, normalized: q, found: false, source: null, entries: [], note: null,
                weblio: `https://ejje.weblio.jp/content/${encodeURIComponent(q)}` };
  if (!q || !/^[a-z][a-z\-']*$/.test(q)) return res;
  const tries = [[q, null], ...inflectionCandidates(q).map((c) => [c, "原形"]), ...derivationCandidates(q).map((c) => [c, "近い語"])];
  for (const [cand, note] of tries) {
    const hits = await ej(cand);
    if (hits.length) return { ...res, found: true, source: "EJDict", headword: hits[0].word, entries: hits, note };
  }
  if (q.includes("-")) {
    for (const part of q.split("-").reverse()) {
      const hits = await ej(part);
      if (hits.length) return { ...res, found: true, source: "EJDict", headword: hits[0].word, entries: hits, note: "部分" };
    }
  }
  const hits = await api(q);
  if (hits.length) return { ...res, found: true, source: "Free Dictionary API（英英）", headword: hits[0].word, entries: hits };
  return res;
}

File: lookup.py
"""英単語を引く。

引く順:
  1. EJDict（<data_root>/dict/ejdict.sqlite。英和）をそのまま
  2. 活用を戻して EJDict（satellites → satellite、measured → measure、flown はそのまま載っている）
  3. 派生語を戻して EJDict（deformable → deform、actuator → actuate）。「近い語」として返す
  4. Free Dictionary API（英英。ネットが要る。つながらなければ飛ばす）
どれでも見つからなければ found=False で、画面は Weblio へのリンクだけ出す。
"""
from __future__ import annotations

import json
import re
import sqlite3
import urllib.parse
import urllib.request
from pathlib import Path

API = "https://api.dictionaryapi.dev/api/v2/entries/en/{}"


def normalize(word: str) -> str:
    w = word.strip().strip(".,;:!?()[]{}\"“”‘’'").lower()
    w = re.sub(r"['’]s$", "", w)
    return w


def inflection_candidates(w: str) -> list[str]:
    """活用形 → 原形の候補（順に試す）。"""
    c = []
    def add(x):
        if len(x) >= 2 and x not in c and x != w:
            c.append(x)
    if w.endswith("ies"):
        add(w[:-3] + "y")
    if w.endswith("es"):
        add(w[:-2])
    if w.endswith("s") and not w.endswith("ss"):
        add(w[:-1])
    if w.endswith("ied"):
        add(w[:-3] + "y")
    if w.endswith("ed"):
        add(w[:-2]); add(w[:-1])
        if len(w) > 4 and w[-3] == w[-4]:
            add(w[:-3])                                  # stopped → stop
    if w.endswith("ing"):
        add(w[:-3]); add(w[:-3] + "e")
        if len(w) > 5 and w[-4] == w[-5]:
            add(w[:-4])                                  # running → run
    if w.endswith("er"):
        add(w[:-2]); add(w[:-1])
    if w.endswith("est"):
        add(w[:-3]); add(w[:-2])
    if w.endswith("ly"):
        add(w[:-2])
    return c


DERIVATIONS = [("ability", ""), ("ability", "e"), ("able", ""), ("able", "e"), ("ible", ""), ("ible", "e"),
               ("ator", "ate"), ("ators", "ate"), ("or", ""), ("ors", ""), ("ation", "ate"), ("ation", "e"),
               ("ations", "ate"), ("ation", ""), ("ment", ""), ("ments", ""), ("ness", ""), ("ity", ""),
               ("ity", "e"), ("ization", "ize"), ("ized", "ize"), ("ically", "ic"), ("ically", "ical")]


def derivation_candidates(w: str) -> list[str]:
    out = []
    for suf, rep in DERIVATIONS:
        if w.endswith(suf) and len(w) - len(suf) >= 3:
            x = w[: -len(suf)] + rep
            if x not in out and x != w:
                out.append(x)
    return out


class Dictionary:
    def __init__(self, data_root: Path, online: bool = True):
        self.path = Path(data_root) / "dict" / "ejdict.sqlite"
        self.online = online

    def available(self) -> bool:
        return self.path.exists()

    def _ej(self, con, w: str) -> list[dict]:
        rows = con.execute("SELECT word, mean FROM words WHERE lower = ? LIMIT 5", (w,)).fetchall()
        return [{"word": r[0], "mean": r[1]} for r in rows]

    def _api(self, w: str) -> list[dict]:
        if not self.online:
            return []
        try:
            req = urllib.request.Request(API.format(urllib.parse.quote(w)), headers={"User-Agent": "paper_reader"})
            with urllib.request.urlopen(req, timeout=5) as r:
                data = json.loads(r.read())
        except Exception:
            return []
        out = []
        for e in data[:1]:
            parts = []
            for m in e.get("meanings", [])[:3]:
                defs = [d["definition"] for d in m.get("definitions", [])[:2]]
                if defs:
                    parts.append(f"({m.get('partOfSpeech', '')}) " + " / ".join(defs))
            if parts:
                out.append({"word": e.get("word", w), "mean": " / ".join(parts)})
        return out

    def lookup(self, word: str) -> dict:
        q = normalize(word)
        res = {"query": word, "normalized": q, "found": False, "source": None, "entries": [], "note": None,
               "weblio": f"https://ejje.weblio.jp/content/{urllib.parse.quote(q)}"}
        if not q or not re.fullmatch(r"[a-z][a-z\-']*", q):
            return res
        if self.available():
            con = sqlite3.connect(self.path)
            try:
                for cand, note in [(q, None)] + [(c, "原形") for c in inflection_candidates(q)] \
                        + [(c, "近い語") for c in derivation_candidates(q)]:
                    hits = self._ej(con, cand)
                    if hits:
                        return {**res, "found": True, "source": "EJDict", "headword": hits[0]["word"],
                                "entries": hits, "note": note}
                if "-" in q:                               # micro-electromechanical → 後半
                    for part in reversed(q.split("-")):
                        hits = self._ej(con, part)
                        if hits:
                            return {**res, "found": True, "source": "EJDict", "headword": hits[0]["word"],
                                    "entries": hits, "note": "部分"}
            finally:
                con.close()
        hits = self._api(q)
        if hits:
            return {**res, "found": True, "source": "Free Dictionary API（英英）", "headword": hits[0]["word"],
                    "entries": hits}
        return res
